utils/s3/client.cc

```diff
@@ -854,6 +864,238 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
+        });
+    }
+
+    future<> complete_multipart_upload() {
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html
+        unsigned parts_xml_len = prepare_multipart_upload_parts(_part_etags);
+        auto req = http::request::make("POST", _client->_host, _object_name);
+        req.query_parameters["uploadId"] = _upload_id;
+        req.write_body("xml", parts_xml_len, [this] (output_stream<char>&& out) -> future<> {
+            return dump_multipart_upload_parts(std::move(out), _part_etags);
+        });
+        // If this request fails, complete_multipart_upload() throws, the upload should then
+        // be aborted in .close() method
+        co_await _client->make_request(std::move(req));
+        _upload_id = ""; // now upload_started() returns false
+    }
+
+    future<> abort_multipart_upload() {
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_AbortMultipartUpload.html
+        s3l.trace("abort multipart upload {}", _upload_id);
+        auto req = http::request::make("DELETE", _client->_host, _object_name);
+        req.query_parameters["uploadId"] = std::exchange(_upload_id, "");
+        co_await _client->make_request(std::move(req), ignore_reply, http::reply::status_type::no_content);
+    }
+
+    // transmit data from input to output in chunks sized up to size
+    static future<> copy_to(input_stream<char>& input,
+                            output_stream<char>& output,
+                            size_t size) {
+        for (;;) {
+            auto buf = co_await input.read_up_to(size);
+            if (buf.empty()) {
+                break;
+            }
+            co_await output.write(buf.get(), buf.size());
+        }
+    }
+
+    future<> upload_part(file f, uint64_t offset, uint64_t part_size) {
+        // upload a part in a multipart upload, see
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html
+        auto mem_units = co_await _client->claim_memory(_transmit_size);
+
+        unsigned part_number = _part_etags.size();
+        _part_etags.emplace_back();
+        auto req = http::request::make("PUT", _client->_host, _object_name);
+        req._headers["Content-Length"] = to_sstring(part_size);
+        req.query_parameters.emplace("partNumber", to_sstring(part_number + 1));
+        req.query_parameters.emplace("uploadId", _upload_id);
+        s3l.trace("PUT part {}, {} bytes (upload id {})", part_number, part_size, _upload_id);
+        req.write_body("bin", part_size, [f=std::move(f), mem_units=std::move(mem_units), offset, part_size] (output_stream<char>&& out_) mutable -> future<> {
+            auto output = std::move(out_);
+            auto input = make_file_input_stream(std::move(f), offset, part_size);
+            std::exception_ptr ex;
+            try {
+                co_await copy_to(input, output, _transmit_size);
```

Comment:
The copy_to() doesn't flush, it means that flush would implicitly happen below when output.close() is called. If it throws there, would the output_stream.close() work correctly and close the stream correctly?

Reply:
ahh, very true. i am adding a `flush()` at the end of `copy_to()`.